calc/rule_scanner.go

```diff
@@ -228,92 +346,172 @@ func ruleToParsedRule(rule *model.Rule) (parsedRule *ParsedRule, allTagOrSels []
+		if p.PortName != "" {
+			namedPorts = append(namedPorts, p.PortName)
+		} else {
+			numericPorts = append(numericPorts, p)
 		}
-		return append(slice, sel)
 	}
+	return
+}
+
+// extractTagsAndSelectors extracts the tag and selector matches from the rule and converts them
+// to selector.Selector objects.  Where it is likely to make the resulting IP sets smaller (or
+// fewer in number), it tries to combine multiple match criteria into a single selector.
+//
+// Returns at most one positive src/dst selector in src/dst.  The named port logic above relies on
+// this.  We still return a slice for those values in order to make it easier to use the utility
+// functions uniformly.
+func extractTagsAndSelectors(rule *model.Rule) (src, dst, notSrc, notDst []selector.Selector) {
+	// Calculate a minimal set of selectors.  We can always combine a positive match on selector
+	// and tag. combineMatchesIfPossible will also try to combine the negative matches into that
+	// single selector, if possible.
+	srcRawSel, notSrcSel, notSrcTag := combineMatchesIfPossible(rule.SrcSelector, rule.SrcTag, rule.NotSrcSelector, rule.NotSrcTag)
+	dstRawSel, notDstSel, notDstTag := combineMatchesIfPossible(rule.DstSelector, rule.DstTag, rule.NotDstSelector, rule.NotDstTag)
 
-	appendSelector := func(slice []selector.Selector, rawSelector string) []selector.Selector {
+	parseAndAppendSelectorIfNonZero := func(slice []selector.Selector, rawSelector string) []selector.Selector {
 		if rawSelector == "" {
 			return slice
 		}
 		sel, err := selector.Parse(rawSelector)
 		if err != nil {
-			// This shouldn't happen because the data should have been validated
-			// further back in the pipeline.
+			// Should have been validated further back in the pipeline.
 			log.WithField("selector", rawSelector).Panic(
 				"Failed to parse selector that should have been validated already.")
 		}
 		return append(slice, sel)
 	}
-
-	src = appendTagSelector(src, rule.SrcTag)
-	src = appendSelector(src, rule.SrcSelector)
-
-	dst = appendTagSelector(dst, rule.DstTag)
-	dst = appendSelector(dst, rule.DstSelector)
-
-	notSrc = appendTagSelector(notSrc, rule.NotSrcTag)
-	notSrc = appendSelector(notSrc, rule.NotSrcSelector)
-
-	notDst = appendTagSelector(notDst, rule.NotDstTag)
-	notDst = appendSelector(notDst, rule.NotDstSelector)
+	src = parseAndAppendSelectorIfNonZero(src, srcRawSel)
+	dst = parseAndAppendSelectorIfNonZero(dst, dstRawSel)
+	notSrc = parseAndAppendSelectorIfNonZero(notSrc, notSrcSel)
+	notSrc = parseAndAppendSelectorIfNonZero(notSrc, tagToSelector(notSrcTag))
+	notDst = parseAndAppendSelectorIfNonZero(notDst, notDstSel)
+	notDst = parseAndAppendSelectorIfNonZero(notDst, tagToSelector(notDstTag))
 
 	return
 }
 
-func selFromTag(tag string) (selector.Selector, error) {
-	return selector.Parse(fmt.Sprintf("has(%s)", tag))
-}
+func combineMatchesIfPossible(positiveSel, positiveTag, notSrcSel, notSrcTag string) (combinedSel, notSrcSelOut, notSrcTagOut string) {
```

Comment:
Not 100% sure, but you have to reuse selectors in lots of different rules to get benefit from the previous approach.

I'm not worried about the evaluation cost, since evaluating selectors is very fast and it does short-circuit evaluation so we may benefit there.

You're right that the worst case of 1 IP set per rule is now more likely but OTOH, it reigns in the old worst case, which was 4, equal or larger IP sets per rule if you don't reuse and selectors or IP sets at all.

I _think_ there are some cases that this makes a lot better too, such as when the negated selector is used for namespace filtering.  You might want to do something like `allow from clients that are not in my namespace`.  Combining the "clients" selector with the negated "in my namespace" selector should be a lot more efficient.

Reply:
Cool, I'm happy with that for now then.  Sounds like we should keep it in mind though, just in case the patterns turn out differently than expected.